qiskit/primitives/backend_estimator_dev.py

```diff
@@ -0,0 +1,619 @@
+        for counts, metadata in zip(counts_bundle, metadata_bundle):
+            paulis = metadata["paulis"]
+            coeffs = metadata["coeffs"]
+            expvals, variances = self._compute_expvals_and_variances(counts, paulis)
+            expval += np.dot(expvals, coeffs)
+            var += np.dot(variances, np.array(coeffs) ** 2)
+        shots = sum(counts_bundle[0].values())  # TODO: not correct -> counts.shots (?)
+        return expval, var, shots
+
+    ################################################################################
+    ## COMPUTATION
+    ################################################################################
+    @classmethod
+    def _compute_expvals_and_variances(
+        cls, counts: Counts, paulis: PauliList
+    ) -> tuple[tuple[float, ...], tuple[float, ...]]:
+        """Return tuples of expvals and variances for input Paulis.
+
+        Note: All non-identity Pauli's are treated as Z-paulis, assuming
+        that basis rotations have been applied to convert them to the
+        diagonal basis.
+        """
+        pairs = tuple(cls._compute_expval_variance_pair(counts, pauli) for pauli in paulis)
+        return tuple(zip(*pairs))
+
+    @classmethod
+    def _compute_expval_variance_pair(cls, counts, pauli):
+        """Return an expval-variance pair for the given counts and pauli.
+
+        Note: All non-identity Pauli's are treated as Z-paulis, assuming
+        that basis rotations have been applied to convert them to the
+        diagonal basis.
+        """
+        shots: int = 0
+        expval: float = 0.0
+        for bitstring, freq in counts.items():
+            coeff = cls._measurement_coefficient(bitstring, pauli)
+            expval += coeff * freq
+            shots += freq
+        if shots == 0:
+            print(counts)
+        expval /= shots
+        variance = 1 - expval**2
+        return expval, variance
+
+    @classmethod
+    def _measurement_coefficient(cls, bitstring: str, pauli: Pauli) -> int:
+        """Compute measurement coefficient from measured bitstring and target Pauli."""
+        measurement = int(bitstring, 2)
+        int_mask = cls._pauli_integer_mask(pauli)
+        return (-1) ** cls._parity_bit(measurement & int_mask, even=True)
+
+    @classmethod
+    def _pauli_integer_mask(cls, pauli: Pauli) -> tuple[int]:
+        """Build integer masks for Pauli.
+
+        This is an integer representation of the binary string with a
+        1 where there are Paulis, and 0 where there are identities.
+        """
+        mask = pauli.z | pauli.x
+        bitstring = cls._bitstring_from_mask(mask, little_endian=True)
+        return int(bitstring, 2)
+
+    @staticmethod
+    def _bitstring_from_mask(mask: Sequence[bool], little_endian: bool = False) -> str:
+        """Return bitstring representation of a one dimensional mask."""
+        if little_endian:
+            mask = reversed(mask)
+        return "0b" + "".join(str(int(b)) for b in mask)
```

Comment:
Hi @ikkoham thanks for the insight, as a matter of fact this implementation is slightly more efficient than the previous one:
```python
from qiskit.quantum_info.operators import SparsePauliOp, Pauli, PauliList
import numpy as np
from qiskit.result import Result, Counts
from qiskit.providers.fake_provider import FakeNairobi, FakeNairobiV2
from collections.abc import Sequence
from functools import reduce
from timeit import timeit

## ORIGINAL
def _paulis2inds(paulis: PauliList) -> list[int]:
    """Convert PauliList to diagonal integers.
    These are integer representations of the binary string with a
    1 where there are Paulis, and 0 where there are identities.
    """
    # Treat Z, X, Y the same
    nonid = paulis.z | paulis.x

    inds = [0] * paulis.size
    # bits are packed into uint8 in little endian
    # e.g., i-th bit corresponds to coefficient 2^i
    packed_vals = np.packbits(nonid, axis=1, bitorder="little")
    for i, vals in enumerate(packed_vals):
        for j, val in enumerate(vals):
            inds[i] += val.item() * (1 << (8 * j))
    return inds

## REFACTOR
_pauli_integer_mask(pauli: Pauli) -> tuple[int]:
    """Build integer masks for Pauli.

    This is an integer representation of the binary string with a
    1 where there are Paulis, and 0 where there are identities.
    """
    pauli_mask = pauli.z | pauli.x
    bitstring = _bitstring_from_mask(pauli_mask, little_endian=True)
    return int(bitstring, 2)

def _bitstring_from_mask(mask: Sequence[bool], little_endian: bool = False) -> str:
    """Return bitstring representation of a one dimensional mask."""
    if little_endian:
        mask = reversed(mask)
    return "0b" + "".join(str(int(b)) for b in mask)

## COMPARISON
size = 10

pauli = Pauli("Z" * size)
puali_list = PauliList([pauli])

before = timeit(lambda: _paulis2inds(puali_list))
after = timeit(lambda: _pauli_integer_mask(pauli))

print("Original: ", before, "(s)")
print("Refactor: ", after, "(s)")
```
```
Original:  5.345644854998682 (s)
Refactor:  5.007759738015011 (s)
```

Furthermore, I realized that the return statement in `_bitstring_from_mask` was not as efficient as it could be so I updated it to:
```python
def _bitstring_from_mask(mask: Sequence[bool], little_endian: bool = False) -> str:
    """Return bitstring representation of a one dimensional mask."""
    if little_endian:
        mask = reversed(mask)
    return "0b" + "".join("1" if b else "0" for b in mask)
```
And now I'm getting significantly better performance on top of an easier to read codebase:
```
Original:  5.3761801840155385 (s)
Refactor:  2.1587249820004217 (s)
```

Reply:
It's for small Paulis. 
Your first code:
```
size = 10
Original:  6.160991474000003 (s)
Refactor:  6.005146135000004 (s)

size = 20
Original:  6.527896039999995 (s)
Refactor:  9.717295750000005 (s)

size = 30
Original:  7.015569922999987 (s)
Refactor:  13.066358127 (s)

size = 40
Original:  7.8512889 (s)
Refactor:  17.304837285000005 (s)
```

Second code:
```
size = 10
Original:  5.949589917000026 (s)
Refactor:  2.487409060999994 (s)

size = 20
Original:  6.544722651999962 (s)
Refactor:  3.123215235000032 (s)

size = 30
Original:  7.3261681060000114 (s)
Refactor:  3.718202363000046 (s)

size = 40
Original:  7.910192152999969 (s)
Refactor:  4.332249084000011 (s)
```
Nice!

However, for square sized PauliList (`puali_list = PauliList([pauli] * size)`)
```
size = 10
Original:  22.434040831000004 (s)
Refactor:  2.4939905560000284 (s) (should be 10 times => 22.49 sec)

size = 20
Original:  50.47780164300002 (s)
Refactor:  3.2489769000000024 (s) (20 terms => 64.97953800000005)
```

It's slower than original.
If abelian grouping is used, the number of terms are large.

Reply:
I run the "square" experiment for `size=40` and this is what I got (after multiplying `after*=size`):
```
Original:  147.28850077901734 (s)
Refactor:  160.48834055894986 (s)
```

Seems negligible _percentage-wise_, especially when put in contrast with the savings that we get at the other end of the spectrum, so imho I'd say the new approach is overall better.

I'll do some more digging to chase down where the inefficiencies come from and what we can do to circumvent them though.

Reply:
After review, I believe the inefficiency comes from repeating `pauli.z | pauli.x` several times instead of doing it once for the entire `PauliList`. This would explain why the overheads are relatively small in the worst case as showcased above.

Reply:
Okay, I wanted to get some proper metrics so I decided to move away from `timeit` and start using a proper profiler. I found several interesting things:
1. `timeit` apparently introduces quite some overhead.
2. The bottleneck in my implementation was in `join`.
3. The bottleneck in the original implementation was in the nested loops.

With all this in mind, I updated the refactor once again to the following hybrid solution:
This is the code that I run:
```python
def _pauli_integer_mask(pauli: Pauli) -> tuple[int]:
    """Build integer masks for Pauli.

    This is an integer representation of the binary string with a
    1 where there are Paulis, and 0 where there are identities.
    """
    pauli_mask: list[bool] = pauli.z | pauli.x
    packed_mask: list[int] = np.packbits(pauli_mask, bitorder="little").tolist()
    return reduce(lambda value, element: (value << 8) + element, packed_mask)
```

Then, I run `cProfile` on a `size=10000` square problem as follows:
```python
from cProfile import Profile
from pstats import Stats, SortKey

size = 10000

pauli = Pauli("Z" * size)
puali_list = PauliList([pauli] * size)

## ORIGINAL
with Profile() as profile:
    before = _paulis2inds(puali_list)
stats = Stats(profile)
stats.sort_stats(SortKey.CUMULATIVE)
stats.dump_stats("original.prof")
print("-" * 20, "ORIGINAL", "-" * 20)
stats.print_stats()

## REFACTOR
with Profile() as profile:
    after = [_pauli_integer_mask(p) for p in puali_list]
stats = Stats(profile)
stats.sort_stats(SortKey.CUMULATIVE)
stats.dump_stats("refactor.prof")
print("-" * 20, "REFACTOR", "-" * 20)
stats.print_stats()

assert before == after
```

Which gives:
```
-------------------- ORIGINAL --------------------
         12500009 function calls in 15.837 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   11.455   11.455   15.837   15.837 <ipython-input-419-0acaac29454f>:1(_paulis2inds)
 12500000    4.371    0.000    4.371    0.000 {method 'item' of 'numpy.generic' objects}
        1    0.000    0.000    0.010    0.010 <__array_function__ internals>:177(packbits)
        1    0.010    0.010    0.010    0.010 {built-in method numpy.core._multiarray_umath.implement_array_function}
      ...
```

```
-------------------- REFACTOR --------------------
         12700006 function calls in 7.622 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.029    0.029    7.622    7.622 <ipython-input-422-4cd7401dfdb9>:20(<listcomp>)
    10000    0.079    0.000    7.485    0.001 <ipython-input-420-2fab7c21caaa>:1(_pauli_integer_mask)
    10000    1.781    0.000    7.270    0.001 {built-in method _functools.reduce}
 12490000    5.490    0.000    5.490    0.000 <ipython-input-420-2fab7c21caaa>:9(<lambda>)
      ...
```

I think it is now safe to say that the new implementation is more performant 🙂 🥳 

Let me know what your thoughts are @ikkoham @t-imamichi, thanks so much for the great reviews!

Reply:
```
def _paulis2inds(paulis: PauliList) -> list[int]:
    nonid = paulis.z | paulis.x
    packed_vals = np.packbits(nonid, axis=1, bitorder="little").astype(object)
    power_uint8 = 1 << (8 * np.arange(packed_vals.shape[1], dtype=object))
    inds = packed_vals @ power_uint8
    return inds.tolist()
```

is faster than your proposal.

Reply:
Nice implementation! I have also improved mine to this:
```python
def _pauli_integer_mask(pauli: Pauli) -> int:
        """Build integer masks for input Pauli.

        This is an integer representation of the binary string with a
        1 where there are Paulis, and 0 where there are identities.
        """
        pauli_mask: np.ndarray[bool] = pauli.z | pauli.x
        packed_mask: list[int] = np.packbits(pauli_mask, bitorder="little").tolist()
        return reduce(lambda value, element: (value << 8) | element, packed_mask)
```

Which now yields:
```
-------------------- IKKO'S NEW --------------------
         10 function calls in 6.090 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    5.978    5.978    5.990    5.990 <ipython-input-583-7e59b4cae6cb>:2(_paulis2inds)
        1    0.000    0.000    0.011    0.011 <__array_function__ internals>:177(packbits)
        1    0.011    0.011    0.011    0.011 {built-in method numpy.core._multiarray_umath.implement_array_function}
        1    0.000    0.000    0.000    0.000 {method 'tolist' of 'numpy.ndarray' objects}
        1    0.000    0.000    0.000    0.000 {built-in method numpy.arange}
      ...
```

```
-------------------- REFACTOR --------------------
         12700006 function calls in 6.483 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.028    0.028    6.653    6.653 <ipython-input-662-4cd7401dfdb9>:20(<listcomp>)
    10000    0.084    0.000    6.520    0.001 <ipython-input-661-341222601da4>:1(_pauli_integer_mask)
    10000    1.790    0.000    6.304    0.001 {built-in method _functools.reduce}
 12490000    4.514    0.000    4.514    0.000 <ipython-input-661-341222601da4>:9(<lambda>)
      ...
```

The reason why your new implementation seems to be a bit faster is that it computes on `PauliList`s at the same time. What I like about doing it on a `Pauli`-by-`Pauli` basis though, is that it opens up the way for cross-observable optimization via caching (i.e. if the operation needs to be repeated for a different observable). What do you think?

Reply:
Quick note as well, in this case the metrics above try to portray rough averages of what I see. Nonetheless, it seems like we are hitting some kind of limit for how fast we can actually do this, because the results seem to fluctuate quite a bit, especially for @ikkoham 's new implementation; where I am getting runtimes oscillating between `~5.7(s)` and `~6.7(s)`.

Reply:
I understand the intention of your benchmark, although I think `abelian_grouping` is most likely to be set to True in actual experiments. Great! Thanks.